rasa/utils/tensorflow/rasa_layers.py

```diff
@@ -0,0 +1,757 @@
+            List[Union[tf.Tensor, tf.SparseTensor]],
+            List[Union[tf.Tensor, tf.SparseTensor]],
+            tf.Tensor,
+            tf.Tensor,
+        ],
+        training: Optional[Union[tf.Tensor, bool]] = None,
+    ) -> Tuple[
+        tf.Tensor,
+        tf.Tensor,
+        Optional[tf.Tensor],
+        Optional[tf.Tensor],
+        Optional[tf.Tensor],
+    ]:
+        """Combine all features for an attribute into one and embed using a transformer.
+
+        Arguments:
+            inputs: Tuple containing:
+                sequence_features: List of 3-D dense or sparse tensors with token-level
+                    features.
+                sentence_features: List of 3-D dense or sparse tensors with sentence-level
+                    features.
+                mask_sequence: a 3-D tensor mask that has 1s at real and 0s at padded
+                    positions corresponding to tokens in `sequence_features`.
+                mask_sequence_sentence: a 3-D tensor mask similar to `mask_sequence` but
+                    having each sequence of 1s longer by 1 to account for sequence 
+                    lengths of sequence- and sentence-level features being combined.
+            training: Python boolean indicating whether the layer should behave in
+                training mode (applying dropout to sparse tensors if applicable) or in
+                inference mode (not applying dropout).
+
+        Returns:
+            outputs: 3-D tensor with all features combined, optionally masked (when doing
+                MLM) and embedded by a transformer.
+            seq_sent_features: 3-D tensor, like `outputs`, but without masking and 
+                transformer applied.
+            token_ids: 3-D tensor with dense token-level features which can serve as
+                unique embeddings/IDs of all the different tokens found in the batch.
+                `None` if not doing MLM.
+            mlm_mask_bool: 3-D tensor mask that has 1s where real tokens in `outputs` 
+                were masked and 0s elsewhere. `None` if not doing MLM.
+            attention_weights: 5-D tensor containing self-attention weights received 
+                from the underlying transformer. `None` if the transformer has 0 layers.
+        """
+        sequence_features = inputs[0]
+        sentence_features = inputs[1]
+        mask_sequence = inputs[2]
+        mask_sequence_sentence = inputs[3]
+
+        # Combine all features (sparse/dense, sequence-/sentence-level) into one tensor
+        seq_sent_features = self._tf_layers["input_layer"](
+            (
+                sequence_features,
+                sentence_features,
+                mask_sequence,
+                mask_sequence_sentence,
+            )
+        )
+
+        seq_sent_features = self._tf_layers["ffnn"](seq_sent_features, training)
+
+        # Produce helper variables for masked language modelling (only in training mode)
+        if self.enables_mlm and training:
```

Comment:
I feel like it is better to abstract it into a separate helper

Reply:
Agree 🙂